Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21752 named inference rule for tensor.select
* #21735 Copy NamedTensorMeta in TensorImpl::copy_tensor_data()
* #21648 Expose torch.empty(sizes, *, names, ...) to Python
* **#21647 Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten**

only

`DimnameList?` is blacklisted in gen_jit_dispatch; the JIT will ignore
`DimnameList?` for the short term. In the long term we can add
first-class support to the JIT for Dimname.

`at::empty(IntArrayRef, DimnameList?, TensorOptions)` is not yet exposed
to python; that requires some extra legwork that is coming soon.

Test Plan
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21647 gh/zou3519/47/head

Differential Revision: [D15804425](https://our.internmc.facebook.com/intern/diff/D15804425)

zou3519 committed:
Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten
only

`DimnameList?` is blacklisted in gen_jit_dispatch; the JIT will ignore
`DimnameList?` for the short term. In the long term we can add
first-class support to the JIT for Dimname.

`at::empty(IntArrayRef, DimnameList?, TensorOptions)` is not yet exposed
to python; that requires some extra legwork that is coming soon.

Test Plan
- New tests [namedtensor ci]

zou3519 committed:
Update on "Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten"

Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten
only

`DimnameList?` is blacklisted in gen_jit_dispatch; the JIT will ignore
`DimnameList?` for the short term. In the long term we can add
first-class support to the JIT for Dimname.

`at::empty(IntArrayRef, DimnameList?, TensorOptions)` is not yet exposed
to python; that requires some extra legwork that is coming soon.

Test Plan
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21647 gh/zou3519/47/head

zou3519 committed:
Update on "Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten"

Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten
only

`DimnameList?` is blacklisted in gen_jit_dispatch; the JIT will ignore
`DimnameList?` for the short term. In the long term we can add
first-class support to the JIT for Dimname.

`at::empty(IntArrayRef, DimnameList?, TensorOptions)` is not yet exposed
to python; that requires some extra legwork that is coming soon.

Test Plan
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21647 gh/zou3519/47/head

nairbv commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -684,6 +684,8 @@
     CPU: _embedding_bag_per_sample_weights_backward_cpu
     CUDA: _embedding_bag_per_sample_weights_backward_cuda
 
+- func: empty(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
Is names=None meaningfully different for purposes of differentiating an unnamed tensor vs just having the default be an empty list? e.g., for a zero-dim tensor, would names=[] mean something different from names=None and would it change any behavior?

Do we need to specify a default value as we do for dtype/layout/etc?

zou3519 replied:
`torch.empty(3, 5, names=None)` creates a TensorImpl with `NamedTensorMetaInterface* = nullptr`.

 `torch.empty(3, 5, names=(None, None))` creates a TensorImpl with `NamedTensorMetaInterface* = <non-null>`.

`at::Tensor::is_named()` returns True if there are any names, otherwise false. It would return false on both of the above two examples.

The plan is for the behavior of the above two tensors to be the same.

zou3519 replied:
No default is necessary because there is an `empty` schema without `Dimname[]? names`.

nairbv replied:
I guess the reason I question it is just that usually when I see list arguments that handle null in a special way, it's often the case that the 'not set' case should just be an empty list. Maybe that's not really applicable here, but we add some overhead for using an optional<list> and theoretically we could just as well represent 'not-set' as an empty list. 

I'm not sure which is clearer to the user in this particular case but null lists always give me pause.

zou3519 replied:
`torch.empty(3, 4, names=[])` throws an error that the number of names does not match the number of dimensions.

zou3519 replied:
Oh, I see what you're saying. I think it makes sense from a usability perspective that:
- `torch.empty(3, 5, names=None)` returns a tensor with no names
- `torch.empty(3, 5, names=[])` errors out, but one can make an argument that it should be the same as the None case.

nairbv replied:
Right, like, we could just entirely get rid of `optional<DimNameList>` and instead just use `DimNameList=()` for unnamed.

nairbv replied:
Arguments can be made for either interface, but we should probably at least think carefully about that tradeoff before we commit to the interface as stable.

zou3519 replied:
Thanks for pointing that out. I haven't thought very much about the problem. Here are some quick thoughts:

For `DimnameList = []`:
- it's easier (and more efficient (but by how much?)) to pass around DimnameList.
- No unwrapping of optional<DimnameList> necessary

For `optional<DimnameList> = None`:
- torch.empty(2, 3, names=[]) feels like an error: user was constructing a names list but then forgot to add names.
- Maintains an invariant that `len(names) == tensor.dim()`.

What other considerations do you think we should take into account?

nairbv replied:
yeah I think that's all. "empty lists over nulls" is 'popular advice' in most languages, though not always applicable.

The cost savings would (I think?) be one address / 8 bytes per named tensor, since there'd be one pointer instead of two? or can the C++ compiler optimize that away somehow?

The invariant basically still looks the same to me, just that it's `names == nullptr || len(names) == tensor.dim()` vs `len(names)==0 || len(names)==tensor.dim()`.

There could be cases where you save a guard, e.g., if you want to `[name for tensor in tensorlist for name in tensor.dimnames()]` to get a full list of all dim names.

Future proofing is probably more significant to consider. I'm not coming up with any particular issues, but you probably have more thoughts on future direction. E.g. if there was some behavioral difference we'd ever want to express in a zero-dim "named" vs "unnamed" tensor, `optional` leaves that path available. If we might someday want to support of an ordered-dict namedtensor constructor, I'm not sure if that would change anything about what interface is ideal.

zou3519 replied:
> The cost savings would (I think?) be one address / 8 bytes per named tensor, since there'd be one pointer instead of two? or can the C++ compiler optimize that away somehow?

Optional does use an extra word, yes. It's not really saving one word per named tensor; it's more like saving one word per operation that uses `optional<DimnameList>`.

> E.g. if there was some behavioral difference we'd ever want to express in a zero-dim "named" vs "unnamed" tensor, optional leaves that path available. 

My plan is not to distinguish between those two; zero-dim tensors (scalars) have no names.

I'll land this patch as-is, but thank you for the discussion, this has been very helpful. I'll revisit this problem when everything is closer to coming together (or if you come up with more ideas/ issues).

nairbv replied:
sounds good, I think it's fine as it is, just wanted to make sure we've considered anything else that could come up.

nairbv commented on aten/src/ATen/test/NamedTensor_test.cpp:
```diff
@@ -43,6 +43,20 @@ TEST(NamedTensorTest, isNamed) {
   ASSERT_TRUE(tensor.is_named());
 }
 
+static bool dimnames_equal(at::DimnameList names, at::DimnameList other) {
```
a dimnamelist is just a std::vector<Dimname> right? does std::vector<Dimname>==() do the same thing?

zou3519 replied:
a `DimnameList` is an `ArrayRef<Dimname>`. I haven't defined `operator==(Dimname,Dimname)` so we can't compare them directly. I'm not sure what I want the semantics of that to be yet.

nairbv replied:
ah, right, I forgot there's some complexity with equality of name vs tag.

zou3519 committed:
Update on "Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten"

Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten
only

`DimnameList?` is blacklisted in gen_jit_dispatch; the JIT will ignore
`DimnameList?` for the short term. In the long term we can add
first-class support to the JIT for Dimname.

`at::empty(IntArrayRef, DimnameList?, TensorOptions)` is not yet exposed
to python; that requires some extra legwork that is coming soon.

Test Plan
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21647 gh/zou3519/47/head

zou3519 committed:
Update on "Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten"

Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten
only

`DimnameList?` is blacklisted in gen_jit_dispatch; the JIT will ignore
`DimnameList?` for the short term. In the long term we can add
first-class support to the JIT for Dimname.

`at::empty(IntArrayRef, DimnameList?, TensorOptions)` is not yet exposed
to python; that requires some extra legwork that is coming soon.

Test Plan
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21647 gh/zou3519/47/head

zou3519 committed:
Update on "Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten"

Implement at::empty(IntArrayRef, DimnameList?, TensorOptions) in aten
only

`DimnameList?` is blacklisted in gen_jit_dispatch; the JIT will ignore
`DimnameList?` for the short term. In the long term we can add
first-class support to the JIT for Dimname.

`at::empty(IntArrayRef, DimnameList?, TensorOptions)` is not yet exposed
to python; that requires some extra legwork that is coming soon.

Test Plan
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 21647 gh/zou3519/47/head
